test/common/integration/xds_integration_test.cc

```diff
@@ -0,0 +1,187 @@
+#include "test/common/integration/xds_integration_test.h"
+#include "xds_integration_test.h"
+
+#include "envoy/config/bootstrap/v3/bootstrap.pb.h"
+#include "envoy/config/cluster/v3/cluster.pb.h"
+
+#include "test/common/grpc/grpc_client_integration.h"
+#include "test/common/integration/base_client_integration_test.h"
+#include "test/test_common/environment.h"
+#include "test/test_common/utility.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+
+using ::testing::AssertionFailure;
+using ::testing::AssertionResult;
+using ::testing::AssertionSuccess;
+
+XdsIntegrationTest::XdsIntegrationTest() : BaseClientIntegrationTest(ipVersion()) {
+  override_builder_config_ = true; // The builder does not yet have RTDS support.
+  expect_dns_ = false;             // TODO(alyssawilk) debug.
+  create_xds_upstream_ = true;
+  sotw_or_delta_ = sotwOrDelta();
+  scheme_ = "https";
+
+  if (sotw_or_delta_ == Grpc::SotwOrDelta::UnifiedSotw ||
+      sotw_or_delta_ == Grpc::SotwOrDelta::UnifiedDelta) {
+    config_helper_.addRuntimeOverride("envoy.reloadable_features.unified_mux", "true");
+  }
+
+  // Set up the basic bootstrap config for xDS.
+  config_helper_.addConfigModifier([this](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+    // The default stats config has overenthusiastic filters.
+    bootstrap.clear_stats_config();
+
+    // Add two clusters by default:
+    //  - base_h2: An HTTP2 cluster with one fake upstream endpoint, for accepting requests from EM.
+    //  - xds_cluster.lyft.com: An xDS management server cluster, with one fake upstream endpoint.
+    bootstrap.mutable_static_resources()->clear_clusters();
+    bootstrap.mutable_static_resources()->add_clusters()->MergeFrom(
+        createSingleEndpointClusterConfig("base_h2"));
+    bootstrap.mutable_static_resources()->add_clusters()->MergeFrom(
+        createSingleEndpointClusterConfig(std::string(XDS_CLUSTER)));
+  });
+
+  // xDS upstream is created separately in the test infra, and there's only one non-xDS cluster.
+  setUpstreamCount(1);
+
+  // Add the Admin config.
+  config_helper_.addConfigModifier([this](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+    bootstrap.mutable_admin()->MergeFrom(adminConfig());
+  });
+  admin_filename_ = TestEnvironment::temporaryPath("admin_address.txt");
+  setAdminAddressPathForTests(admin_filename_);
+}
+
+Network::Address::IpVersion XdsIntegrationTest::ipVersion() const {
+  return std::get<0>(GetParam());
+}
+
+Grpc::ClientType XdsIntegrationTest::clientType() const { return std::get<1>(GetParam()); }
+
+Grpc::SotwOrDelta XdsIntegrationTest::sotwOrDelta() const { return std::get<2>(GetParam()); }
+
+std::string XdsIntegrationTest::loopbackAddr() const {
```

Comment:
Network::Utility::getLocalAddress(ipVersion()) ?

Reply:
I think getLocalAddress actually(ipVersion) will handle the v4/v6 if for you

Reply:
Changed to Network::Test::getLoopbackAddressString()